### PR TITLE
stdenv: update ARM bootstrap tarballs

### PR DESCRIPTION
**Pull request by lopsided98**
###### Motivation for this change

Updates the 32-bit ARM bootstrap tarballs to the latest cross-compiled versions built by Hydra in the [`cross-trunk`](https://hydra.nixos.org/jobset/nixpkgs/cross-trunk) jobset. This is required because glibc 2.30 requires at least GCC6 (see https://github.com/NixOS/nixpkgs/pull/66528).

This update is based on the previous one in https://github.com/NixOS/nixpkgs/commit/28f87e414186e2f16a094ac354dfe079965e8fe5.

I have not tested this at all yet, but @thefloweringash [reported armv7l build failures](https://logs.nix.samueldr.com/nixos-aarch64/2020-02-11#3059537;).

The tools are currently hosted on my server, but will need to be moved to tarballs.nixos.org before this is merged.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @samueldr @thefloweringash @vcunat

### CODE REVIEW

**Comment by samueldr** [comment]
I figure those are the [`.dist` jobs](https://hydra.nixos.org/jobset/nixpkgs/cross-trunk/jobs-tab?filter=.dist), right?

Assuming one doesn't trust your tarballs, they could use those instead, right?

The next step might be to somewhat automatize this step of taking the latest bootstrap tarballs? Any issues with the idea? I'm thinking a maintainer script that could be run manually to produce the update and a list of tarballs to download from hydra, with appropriate names.

**Comment by lopsided98** [comment]
I downloaded them from the individual jobs (`bootstrapFiles.busybox` and `bootstrapFiles.bootstrapTools`) because I didn't know about the `dist` job. I will update the URLs to download directly from Hydra.

**Comment by lopsided98** [comment]
I updated the URLs to point to Hydra. The busybox `sha256` does not match if I use the value obtained from Hydra or manual `sha256sum`. I fixed the armv7l one by TOFU but the others are still incorrect.

gcc in the bootstrap tarball is nonfunctional because it is missing libssp.

**Comment by thefloweringash** [comment]
This seems to get a little further, but fails on xz:
```
libtool: link: gcc -fvisibility=hidden -Wall -Wextra -Wvla -Wformat=2 -Winit-self -Wmissing-include-dirs -Wstrict-aliasing -Wfloat-equal -Wundef -Wshadow -Wpointer-arith -Wbad-function-cast -Wwrite-strings -Wlogical-op -Waggregate-return -Wstrict-prototypes -Wold-style-definition -Wmissing-prototypes -Wmissing-declarations -Wmissing-noreturn -Wredundant-decls -g -O2 -o .libs/xzdec xzdec-xzdec.o xzdec-tuklib_progname.o xzdec-tuklib_exit.o  ../../src/liblzma/.libs/liblzma.so -Wl,-rpath -Wl,/nix/store/njjhdzw71igfdy9zibyij8nv8nqgapac-xz-5.2.4/lib
/nix/store/xhkjbgb64krykh5mij8zryvswaayr9j8-bootstrap-tools/bin/ld: ../../src/liblzma/.libs/liblzma.so: undefined reference to `pthread_sigmask'
/nix/store/xhkjbgb64krykh5mij8zryvswaayr9j8-bootstrap-tools/bin/ld: ../../src/liblzma/.libs/liblzma.so: undefined reference to `pthread_join'
/nix/store/xhkjbgb64krykh5mij8zryvswaayr9j8-bootstrap-tools/bin/ld: ../../src/liblzma/.libs/liblzma.so: undefined reference to `pthread_create'
```

**Comment by vcunat** [comment]
> The next step might be to somewhat automatize this step [...]

In any case I believe updates of bootstrap tarball (or busybox) should be rather rare events, hopefully years apart.  Right now the various failures and fixups seem to need much more work than the automatizable parts.

**Comment by lopsided98** [comment]
For me it failed building GCC. I ran into the the same error as on i686 in https://github.com/NixOS/nixpkgs/pull/66528#issuecomment-582332174. I am currently testing to see if disabling LTO works around the problem.

**Comment by samueldr** [comment]
> I believe updates of bootstrap tarball (or busybox) should be rather rare events

Yes, I agree, hence "somewhat" automatize. My main thought here is that each time it happens it's years apart, people come and go, people forget what needs to happen to do this properly.

**Comment by vcunat** [comment]
For that it's enough to write some docs/howto :-)

**Comment by lopsided98** [comment]
I can't reproduce the xz failure. I was able to build armv7l stdenv except for a failing `allowedRequisites` check because `expand-response-params` has bootstrap-tools in its rpath.

**Comment by thefloweringash** [comment]
I'm testing this again, but there's a bit being left unsaid in this thread so I'm not sure I'm testing the right thing. Check my working and see where I'm going wrong.

First off, this branch doesn't evaluate as is due to infinite recursion.

```
$ git fetch origin pull/79793/head && git checkout FETCH_HEAD
[...]
$ nix-build pkgs/top-level/release-cross.nix -A bootstrapTools.armv7l.dist      
error: infinite recursion encountered, at undefined position
(use '--show-trace' to show detailed location information)
```

So I'm testing against the merge, which as of writing is:

```
$ git fetch origin pull/79793/merge && git checkout FETCH_HEAD
[...]
HEAD is now at 16c7ca5dd10 Merge 1f1b31c05afe88b45519d93a46b0bc9eafb5e490 into 816e214c1838d7e7265cdf1ffa361bfac852b38d

$ nix-build pkgs/top-level/release-cross.nix -A bootstrapTools.armv7l.dist
/nix/store/n3l34vzl1hdi5a7vs4rii4qgfh98zi8x-stdenv-bootstrap-tools-armv7l-unknown-linux-gnueabihf
```

I then plug the results into `pkgs/stdenv/linux/bootstrap-files/armv7l.nix` and `nix-build --argstr system armv7l-linux -A hello`. This builds for a while, then stops after
```
/nix/store/lmss9iwplnwcngv4xvhp4nsl9lkwh70n-bootstrap-tools/bin/ld: ../../src/liblzma/.libs/liblzma.so: undefined reference to `pthread_sigmask'
/nix/store/lmss9iwplnwcngv4xvhp4nsl9lkwh70n-bootstrap-tools/bin/ld: ../../src/liblzma/.libs/liblzma.so: undefined reference to `pthread_join'
/nix/store/lmss9iwplnwcngv4xvhp4nsl9lkwh70n-bootstrap-tools/bin/ld: ../../src/liblzma/.libs/liblzma.so: undefined reference to `pthread_create'
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:501: xzdec] Error 1
make[3]: Leaving directory '/build/xz-5.2.4/src/xzdec'
make[2]: *** [Makefile:420: all-recursive] Error 1
make[2]: Leaving directory '/build/xz-5.2.4/src'
make[1]: *** [Makefile:613: all-recursive] Error 1
make[1]: Leaving directory '/build/xz-5.2.4'
make: *** [Makefile:480: all] Error 2
builder for '/nix/store/93qgcgmsyk5lidq8xx37488677ics6sj-xz-5.2.4.drv' failed with exit code 2

```

To reduce error I made a [quick script][] to glue the two halves together.

[quick script]: https://gist.github.com/thefloweringash/ec9511da8c7f41f903eb530a68f7e30f

**Comment by lopsided98** [comment]
My testing was done by applying https://github.com/NixOS/nixpkgs/pull/79793/commits/7e7bbcbc29bc689c58a8430da3529ad5c70fa0d0 on top of ac241fb7a570d6cf81d229ad22a8889602639160 (the latest commit that Hydra had finished building bootstrap tools for at the time I started testing) in order to build the bootstrap tools.

I was planning to split 7e7bbcbc29bc689c58a8430da3529ad5c70fa0d0 into its own PR and then let Hydra build a new version of the bootstrap tools. Based on your error, it looks like now some other change has broken the bootstrap tools again.

**Comment by lopsided98** [comment]
I fixed the issue that was causing to the `xz` build to fail. The `unpack-bootstrap-tools.sh` script was patching a Nix store path in the `libc.so` and `libpthread.so` linker scripts. On 32-bit ARM, `libpthread.so` is no longer a linker script but a shared library, so this corrupted the file.

AFAICT, the linker scripts no longer contain any Nix store paths on any platform, so I just removed the patching commands.

I also created #80497, which will allow Hydra to build working bootstrap tarballs for ARM.

**Comment by thefloweringash** [comment]
I've successfully built `hello` using these branches. The changes seem sufficient. 👍

**Comment by lopsided98** [comment]
Now that the bootstrap tools have glibc 2.30, LTO works again but will probably break with the next glibc update. I'm not sure if I should keep it disabled, or wait and see if it becomes a problem next time glibc is updated.

**Comment by lopsided98** [comment]
Also, I have tested building armv6l stdenv on an armv7l machine (using https://github.com/lopsided98/nixpkgs/commit/001be94c649a75decde8733da4d1289d85072898, which is how I normally build for armv6l).

I have not tested armv5tel, but this PR can't make it any more broken than it currently is.

**Comment by vcunat** [comment]
PR #79391 for glibc 2.31 is there already, if you want to try now.

**Comment by lopsided98** [comment]
`stdenv` builds successfully with glibc 2.31 and LTO enabled, so I removed the commit that disables it on ARM.

File: pkgs/stdenv/linux/bootstrap-files/armv7l.nix [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
+    sha256 = "566a469dac214b31e4abdb0a91d32550bab1be5858d329e1b6074eef05370ca3";
     executable = true;
   };
 
   bootstrapTools = import <nix/fetchurl.nix> {
-    url = http://nixos-arm.dezgeg.me/bootstrap-2017-04-13-1f32d4b4/armv7l/bootstrap-tools.tar.xz;
-    sha256 = "05ayki2kak3i5lw97qidd5h9jv00dmlhx9h7l771bj331yamyqdn";
+    url = https://hydra.nixos.org/build/112609103/download/1/bootstrap-tools.tar.xz;
+    sha256 = "79fa2d7722aeb856c7c9b62a3fd64b6d261fd6f6bcbac486f0a2a7d823210550";
   };
```
Comment (grahamc) [inline pkgs/stdenv/linux/bootstrap-files/armv7l.nix]:
```suggestion
  busybox = import <nix/fetchurl.nix> {
    # from job: https://hydra.nixos.org/job/nixpkgs/cross-trunk/bootstrapTools.armv7l.dist/latest
    # from build: https://hydra.nixos.org/build/113692702
    url = https://tarballs.nixos.org/sha256/7a7a0266d4687eb1b44496d4d185b9ffa9b1c6260c5df1c40a4999c612312bbd;
    sha256 = "7a7a0266d4687eb1b44496d4d185b9ffa9b1c6260c5df1c40a4999c612312bbd";
    executable = true;
  }

  bootstrapTools = import <nix/fetchurl.nix> {
    # from job: https://hydra.nixos.org/job/nixpkgs/cross-trunk/bootstrapTools.armv7l.dist/latest
    # from build: https://hydra.nixos.org/build/113692702
    url = https://tarballs.nixos.org/sha256/593772a3524b9c37527c0c23375fe8d792d5401fbd0f703940e70df470cc129e;
    sha256 = "593772a3524b9c37527c0c23375fe8d792d5401fbd0f703940e70df470cc129e";
  }
```

File: pkgs/stdenv/linux/bootstrap-files/armv6l.nix [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
+    sha256 = "e6f6aecb675e924a96516f4379445dd2c0ba8b9c438fbfbaa2dc14ccce2802e0";
     executable = true;
   };
 
   bootstrapTools = import <nix/fetchurl.nix> {
-    url = http://nixos-arm.dezgeg.me/bootstrap-2017-04-13-1f32d4b4/armv6l/bootstrap-tools.tar.xz;
-    sha256 = "1gg2q3sw81vi65g1gmpvx0nnd6hxb76vlz73wfp292m90z1mym7f";
+    url = https://hydra.nixos.org/build/112609441/download/1/bootstrap-tools.tar.xz;
+    sha256 = "7a3f20def1a17ebf0edb5a92c403558429bcc2ac3d931b5e1bd88606cb217778";
   };
```
Comment (grahamc) [inline pkgs/stdenv/linux/bootstrap-files/armv6l.nix]:
```suggestion
  busybox = import <nix/fetchurl.nix> {
    # from job: https://hydra.nixos.org/job/nixpkgs/cross-trunk/bootstrapTools.armv6l.dist/latest
    # from build: https://hydra.nixos.org/build/113692702
    url = https://tarballs.nixos.org/sha256/d96c7656bf43e9ab24949a0b7f6c7c4a4daf1f89ce10ea6e86aa645cbff65bce;
    sha256 = "d96c7656bf43e9ab24949a0b7f6c7c4a4daf1f89ce10ea6e86aa645cbff65bce";
    executable = true;
  }

  bootstrapTools = import <nix/fetchurl.nix> {
    # from job: https://hydra.nixos.org/job/nixpkgs/cross-trunk/bootstrapTools.armv6l.dist/latest
    # from build: https://hydra.nixos.org/build/113692702
    url = https://tarballs.nixos.org/sha256/956bb4870a68ade2e842b19a2b61f887f2b40e9ca25874592fe29abfc1a107e4;
    sha256 = "956bb4870a68ade2e842b19a2b61f887f2b40e9ca25874592fe29abfc1a107e4";
  }
```

File: pkgs/stdenv/linux/bootstrap-files/armv5tel.nix [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
+    sha256 = "0dc5471dc6a5f69ad98eb7445f51a61e88aa5792d7a677025bf012bdb513b763";
     executable = true;
   };
 
   bootstrapTools = import <nix/fetchurl.nix> {
-    url = http://nixos-arm.dezgeg.me/bootstrap-2017-04-13-1f32d4b4/armv5tel/bootstrap-tools.tar.xz;
-    sha256 = "0fhiy9l3mbmlhpkby31c2s63bhjiqx25qqr3wdp8cb7fxz8ayx2f";
+    url = https://hydra.nixos.org/build/112609163/download/1/bootstrap-tools.tar.xz;
+    sha256 = "ca0564eca4eb944649ce10ec70859640427bf2241243af62812b163176487e02";
   };
```
Comment (grahamc) [inline pkgs/stdenv/linux/bootstrap-files/armv5tel.nix]:
```suggestion
  busybox = import <nix/fetchurl.nix> {
    # from job: https://hydra.nixos.org/job/nixpkgs/cross-trunk/bootstrapTools.armv5tel.dist/latest
    # from build: https://hydra.nixos.org/build/113692702
    url = https://tarballs.nixos.org/sha256/3091d8e22778d023d51c50526ac4911b617db38b7f092fa3332c1d3ff5bd53b7;
    sha256 = "3091d8e22778d023d51c50526ac4911b617db38b7f092fa3332c1d3ff5bd53b7";
    executable = true;
  }

  bootstrapTools = import <nix/fetchurl.nix> {
    # from job: https://hydra.nixos.org/job/nixpkgs/cross-trunk/bootstrapTools.armv5tel.dist/latest
    # from build: https://hydra.nixos.org/build/113692702
    url = https://tarballs.nixos.org/sha256/c648fbc9eab9ff58e4d088c6ffe9d7cf6c807115c351a4bb7d1a040db0ecb2ec;
    sha256 = "c648fbc9eab9ff58e4d088c6ffe9d7cf6c807115c351a4bb7d1a040db0ecb2ec";
  }
```

**Review by grahamc — COMMENTED** [review]
_deleted_ we need to regenerate the bootstrap tarballs after #79793.

**Review by LnL7 — APPROVED** [review]

**Comment by lopsided98** [comment]
@FRidh This should not have been merged.

**Comment by grahamc** [comment]
The problem part is the bootstrap URLs, right? If so, we can follow-up with an update.

**Comment by lopsided98** [comment]
Yes, this shouldn't break anything any more than it already is.

Just so everything is clear, these are the required next steps:

1. Merge #80497 
2. Wait for Hydra to build new bootstrap tarballs
3. Upload new bootstrap tarballs to tarball.nixos.org
4. Create a new PR that updates nixpkgs to use the new tarballs

**Pull request merged**